Use core atomic types instead of intrinsics on non-MSP430 targets

taiki-e opened the pull request:
Since https://github.com/rust-lang/rust/pull/97423, build on non-MSP430 targets is failing.



<details><summary>log</summary>

```
error[E0425]: cannot find function `atomic_load` in module `core::intrinsics`
   --> src/lib.rs:684:37
    |
684 |                   ::core::intrinsics::atomic_load(dst)
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
720 | / atomic_int! {
721 | |     i8 AtomicI8 ATOMIC_I8_INIT ".b"
722 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xadd` in module `core::intrinsics`
   --> src/lib.rs:689:37
    |
689 |                   ::core::intrinsics::atomic_xadd(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
720 | / atomic_int! {
721 | |     i8 AtomicI8 ATOMIC_I8_INIT ".b"
722 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xsub` in module `core::intrinsics`
   --> src/lib.rs:694:37
    |
694 |                   ::core::intrinsics::atomic_xsub(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
720 | / atomic_int! {
721 | |     i8 AtomicI8 ATOMIC_I8_INIT ".b"
722 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:699:37
    |
699 |                   ::core::intrinsics::atomic_and(dst, val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
720 | / atomic_int! {
721 | |     i8 AtomicI8 ATOMIC_I8_INIT ".b"
722 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:704:37
    |
704 |                   ::core::intrinsics::atomic_and(dst, !val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
720 | / atomic_int! {
721 | |     i8 AtomicI8 ATOMIC_I8_INIT ".b"
722 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_or` in module `core::intrinsics`
   --> src/lib.rs:709:37
    |
709 |                   ::core::intrinsics::atomic_or(dst, val);
    |                                       ^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
720 | / atomic_int! {
721 | |     i8 AtomicI8 ATOMIC_I8_INIT ".b"
722 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xor` in module `core::intrinsics`
   --> src/lib.rs:714:37
    |
714 |                   ::core::intrinsics::atomic_xor(dst, val);
    |                                       ^^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
720 | / atomic_int! {
721 | |     i8 AtomicI8 ATOMIC_I8_INIT ".b"
722 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_load` in module `core::intrinsics`
   --> src/lib.rs:684:37
    |
684 |                   ::core::intrinsics::atomic_load(dst)
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
724 | / atomic_int! {
725 | |     u8 AtomicU8 ATOMIC_U8_INIT ".b"
726 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xadd` in module `core::intrinsics`
   --> src/lib.rs:689:37
    |
689 |                   ::core::intrinsics::atomic_xadd(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
724 | / atomic_int! {
725 | |     u8 AtomicU8 ATOMIC_U8_INIT ".b"
726 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xsub` in module `core::intrinsics`
   --> src/lib.rs:694:37
    |
694 |                   ::core::intrinsics::atomic_xsub(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
724 | / atomic_int! {
725 | |     u8 AtomicU8 ATOMIC_U8_INIT ".b"
726 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:699:37
    |
699 |                   ::core::intrinsics::atomic_and(dst, val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
724 | / atomic_int! {
725 | |     u8 AtomicU8 ATOMIC_U8_INIT ".b"
726 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:704:37
    |
704 |                   ::core::intrinsics::atomic_and(dst, !val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
724 | / atomic_int! {
725 | |     u8 AtomicU8 ATOMIC_U8_INIT ".b"
726 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_or` in module `core::intrinsics`
   --> src/lib.rs:709:37
    |
709 |                   ::core::intrinsics::atomic_or(dst, val);
    |                                       ^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
724 | / atomic_int! {
725 | |     u8 AtomicU8 ATOMIC_U8_INIT ".b"
726 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xor` in module `core::intrinsics`
   --> src/lib.rs:714:37
    |
714 |                   ::core::intrinsics::atomic_xor(dst, val);
    |                                       ^^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
724 | / atomic_int! {
725 | |     u8 AtomicU8 ATOMIC_U8_INIT ".b"
726 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_load` in module `core::intrinsics`
   --> src/lib.rs:684:37
    |
684 |                   ::core::intrinsics::atomic_load(dst)
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
728 | / atomic_int! {
729 | |     i16 AtomicI16 ATOMIC_I16_INIT ".w"
730 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xadd` in module `core::intrinsics`
   --> src/lib.rs:689:37
    |
689 |                   ::core::intrinsics::atomic_xadd(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
728 | / atomic_int! {
729 | |     i16 AtomicI16 ATOMIC_I16_INIT ".w"
730 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xsub` in module `core::intrinsics`
   --> src/lib.rs:694:37
    |
694 |                   ::core::intrinsics::atomic_xsub(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
728 | / atomic_int! {
729 | |     i16 AtomicI16 ATOMIC_I16_INIT ".w"
730 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:699:37
    |
699 |                   ::core::intrinsics::atomic_and(dst, val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
728 | / atomic_int! {
729 | |     i16 AtomicI16 ATOMIC_I16_INIT ".w"
730 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:704:37
    |
704 |                   ::core::intrinsics::atomic_and(dst, !val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
728 | / atomic_int! {
729 | |     i16 AtomicI16 ATOMIC_I16_INIT ".w"
730 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_or` in module `core::intrinsics`
   --> src/lib.rs:709:37
    |
709 |                   ::core::intrinsics::atomic_or(dst, val);
    |                                       ^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
728 | / atomic_int! {
729 | |     i16 AtomicI16 ATOMIC_I16_INIT ".w"
730 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xor` in module `core::intrinsics`
   --> src/lib.rs:714:37
    |
714 |                   ::core::intrinsics::atomic_xor(dst, val);
    |                                       ^^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
728 | / atomic_int! {
729 | |     i16 AtomicI16 ATOMIC_I16_INIT ".w"
730 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_load` in module `core::intrinsics`
   --> src/lib.rs:684:37
    |
684 |                   ::core::intrinsics::atomic_load(dst)
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
732 | / atomic_int! {
733 | |     u16 AtomicU16 ATOMIC_U16_INIT ".w"
734 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xadd` in module `core::intrinsics`
   --> src/lib.rs:689:37
    |
689 |                   ::core::intrinsics::atomic_xadd(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
732 | / atomic_int! {
733 | |     u16 AtomicU16 ATOMIC_U16_INIT ".w"
734 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xsub` in module `core::intrinsics`
   --> src/lib.rs:694:37
    |
694 |                   ::core::intrinsics::atomic_xsub(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
732 | / atomic_int! {
733 | |     u16 AtomicU16 ATOMIC_U16_INIT ".w"
734 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:699:37
    |
699 |                   ::core::intrinsics::atomic_and(dst, val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
732 | / atomic_int! {
733 | |     u16 AtomicU16 ATOMIC_U16_INIT ".w"
734 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:704:37
    |
704 |                   ::core::intrinsics::atomic_and(dst, !val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
732 | / atomic_int! {
733 | |     u16 AtomicU16 ATOMIC_U16_INIT ".w"
734 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_or` in module `core::intrinsics`
   --> src/lib.rs:709:37
    |
709 |                   ::core::intrinsics::atomic_or(dst, val);
    |                                       ^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
732 | / atomic_int! {
733 | |     u16 AtomicU16 ATOMIC_U16_INIT ".w"
734 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xor` in module `core::intrinsics`
   --> src/lib.rs:714:37
    |
714 |                   ::core::intrinsics::atomic_xor(dst, val);
    |                                       ^^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
732 | / atomic_int! {
733 | |     u16 AtomicU16 ATOMIC_U16_INIT ".w"
734 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_load` in module `core::intrinsics`
   --> src/lib.rs:684:37
    |
684 |                   ::core::intrinsics::atomic_load(dst)
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
736 | / atomic_int! {
737 | |     isize AtomicIsize ATOMIC_ISIZE_INIT ".w"
738 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xadd` in module `core::intrinsics`
   --> src/lib.rs:689:37
    |
689 |                   ::core::intrinsics::atomic_xadd(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
736 | / atomic_int! {
737 | |     isize AtomicIsize ATOMIC_ISIZE_INIT ".w"
738 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xsub` in module `core::intrinsics`
   --> src/lib.rs:694:37
    |
694 |                   ::core::intrinsics::atomic_xsub(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
736 | / atomic_int! {
737 | |     isize AtomicIsize ATOMIC_ISIZE_INIT ".w"
738 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:699:37
    |
699 |                   ::core::intrinsics::atomic_and(dst, val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
736 | / atomic_int! {
737 | |     isize AtomicIsize ATOMIC_ISIZE_INIT ".w"
738 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:704:37
    |
704 |                   ::core::intrinsics::atomic_and(dst, !val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
736 | / atomic_int! {
737 | |     isize AtomicIsize ATOMIC_ISIZE_INIT ".w"
738 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_or` in module `core::intrinsics`
   --> src/lib.rs:709:37
    |
709 |                   ::core::intrinsics::atomic_or(dst, val);
    |                                       ^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
736 | / atomic_int! {
737 | |     isize AtomicIsize ATOMIC_ISIZE_INIT ".w"
738 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xor` in module `core::intrinsics`
   --> src/lib.rs:714:37
    |
714 |                   ::core::intrinsics::atomic_xor(dst, val);
    |                                       ^^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
736 | / atomic_int! {
737 | |     isize AtomicIsize ATOMIC_ISIZE_INIT ".w"
738 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_load` in module `core::intrinsics`
   --> src/lib.rs:684:37
    |
684 |                   ::core::intrinsics::atomic_load(dst)
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
740 | / atomic_int! {
741 | |     usize AtomicUsize ATOMIC_USIZE_INIT ".w"
742 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xadd` in module `core::intrinsics`
   --> src/lib.rs:689:37
    |
689 |                   ::core::intrinsics::atomic_xadd(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
740 | / atomic_int! {
741 | |     usize AtomicUsize ATOMIC_USIZE_INIT ".w"
742 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xsub` in module `core::intrinsics`
   --> src/lib.rs:694:37
    |
694 |                   ::core::intrinsics::atomic_xsub(dst, val);
    |                                       ^^^^^^^^^^^ not found in `core::intrinsics`
...
740 | / atomic_int! {
741 | |     usize AtomicUsize ATOMIC_USIZE_INIT ".w"
742 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:699:37
    |
699 |                   ::core::intrinsics::atomic_and(dst, val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
740 | / atomic_int! {
741 | |     usize AtomicUsize ATOMIC_USIZE_INIT ".w"
742 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_and` in module `core::intrinsics`
   --> src/lib.rs:704:37
    |
704 |                   ::core::intrinsics::atomic_and(dst, !val);
    |                                       ^^^^^^^^^^ not found in `core::intrinsics`
...
740 | / atomic_int! {
741 | |     usize AtomicUsize ATOMIC_USIZE_INIT ".w"
742 | | }
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_or` in module `core::intrinsics`
   --> src/lib.rs:709:37
    |
709 |                   ::core::intrinsics::atomic_or(dst, val);
    |                                       ^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
740 | / atomic_int! {
741 | |     usize AtomicUsize ATOMIC_USIZE_INIT ".w"
742 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `atomic_xor` in module `core::intrinsics`
   --> src/lib.rs:714:37
    |
714 |                   ::core::intrinsics::atomic_xor(dst, val);
    |                                       ^^^^^^^^^^ help: a function with a similar name exists: `atomic_store`
...
740 | / atomic_int! {
741 | |     usize AtomicUsize ATOMIC_USIZE_INIT ".w"
742 | | }
    | |_- in this macro invocation
    |
   ::: /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:523:5
    |
523 |       pub fn atomic_store_seqcst<T: Copy>(dst: *mut T, val: T);
    |       -------------------------------------------------------- similarly named function `atomic_store` defined here
    |
    = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)
```

</details>

taiki-e commented:
Btw, an alternative idea is using core atomic types instead of intrinsics on non-MSP430 targets: https://github.com/taiki-e/msp430-atomic/commit/a5512a988db5cef9d28ff354feadee8f5f92943d

cr1901 commented:
I don't remember the rationale for making the API function names reflect the intrinsics rather than "the API provided by `core::sync::atomic`.

cc: @pftbest Do you object if the API is updated (_in a breaking change_) to reflect `core::sync::atomic` function names rather than `core::intrinsics` functions?

pftbest commented:
Hello @cr1901, The idea of this crate was to keep public API exactly the same as in std to simply porting code written for another platforms. The internal implementation (ie intrinsics, inline asm, atomic types, etc) doesn't really matter as long as public API behaves as expected. I am ok with any internal changes to that effect.

cr1901 commented:
@pftbest

>The idea of this crate was to keep public API exactly the same as in std to simply porting code written for another platforms

I'm a little bit confused. Right now, the names in the public API of this crate reflect `core::instrinics`, not `core::sync::atomic`. I'm not sure if `core::intrinsics` is ever meant to be stabilized. Was this intentional at the time?

Additionally, msp430 RMW insns [are not sufficient](https://sourceforge.net/p/mspgcc/mailman/message/29912680/) for implementing the return value part of the functions provided by `core::sync::atomic`, only the actual atomic operation. Actually returning the value in a register would require a second memory access. So this crate _already_ didn't have the same API as `core::sync::atomic` _or_ `core::intrinsics`.

pftbest commented:
@cr1901 Sorry, I guess I did forget this part. You are right it's not possible to have a matching API. I'm not sure why I went with the intrinsics API at the time. Following intrinsics API doesn't really make sense because it's unstable by definition and it's not used in practice. Updating API to match `core::sync::atomic` looks like a better idea. Please go ahead.

cr1901 commented:
>Updating API to match core::sync::atomic looks like a better idea. Please go ahead.

@pftbest I will hold onto this as an alternate _alternate_ idea. I think I want to deprecate this crate ~once `fence`/`compiler_fence` is actually working for MSP430 LLVM (it doesn't lower properly right now).~ EDIT: I don't think this is relevant to msp430 and we can deprecate whenever:

* `portable-atomic` reuses the inline asm versions from this crate to provide msp430 1/8/16-bit load and store, and those implicitly introduce a `SeqCst` compiler fence (hardware fence is irrelevant for single core).
* `atomic-polyfill` for everything- and `portable-atomic` for anything not in the previous bullet- uses critical sections, the beginning and end of which are implemented in inline asm and implicitly introduce `SeqCst` compiler fences around the closure that must run with interrupts disabled.

@taiki-e Do you mind if we use your [alternate idea](https://github.com/taiki-e/msp430-atomic/commit/a5512a988db5cef9d28ff354feadee8f5f92943d) for this PR? I think short-to-medium term plans are to deprecate this crate and either use your [portable-atomic](https://github.com/taiki-e/portable-atomic) crate (looks fine, although the scary `--cfg` option is not necessary for msp430- _all_ msp430s are single-core)  and/or add msp430 support to the [atomic-polyfill](https://github.com/embassy-rs/atomic-polyfill) crate (which `rust-embedded` seems to be using more).

taiki-e commented:
@cr1901 

> Do you mind if we use your [alternate idea](https://github.com/taiki-e/msp430-atomic/commit/a5512a988db5cef9d28ff354feadee8f5f92943d) for this PR?

Updated to use it. 

> I think short-to-medium term plans are to deprecate this crate and either use your [portable-atomic](https://github.com/taiki-e/portable-atomic) crate (looks fine, although the scary `--cfg` option is not necessary for msp430- _all_ msp430s are single-core) and/or add msp430 support to the [atomic-polyfill](https://github.com/embassy-rs/atomic-polyfill) crate (which `rust-embedded` seems to be using more).

Sounds good to me. For `portable-atomic`, I've updated it to remove the need for the unsafe cfg for msp430. (https://github.com/taiki-e/portable-atomic/pull/31)
